plugins/samples/hmac_authcookie/plugin.cc

```diff
@@ -0,0 +1,92 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START serviceextensions_plugin_hmac_authcookie]
+#include <openssl/hmac.h>
+
+#include <iomanip>
+#include <sstream>
+#include <string>
+
+#include "absl/strings/str_split.h"
+#include "proxy_wasm_intrinsics.h"
+
+// Replace with your desired secret key.
+const std::string kSecretKey = "your_secret_key";
+
+class MyHttpContext : public Context {
+ public:
+  explicit MyHttpContext(uint32_t id, RootContext* root) : Context(id, root) {}
+
+  FilterHeadersStatus onRequestHeaders(uint32_t headers,
+                                       bool end_of_stream) override {
+    const auto token = getTokenFromCookie();
+    if (!token.has_value()) {
+      LOG_INFO("Access forbidden - missing HMAC cookie.");
+      sendLocalResponse(403, "", "Access forbidden - missing HMAC cookie.\n",
+                        {});
+      return FilterHeadersStatus::ContinueAndEndStream;
+    }
+
+    const auto path = getRequestHeader(":path")->toString();
+    if (computeHmacSignature(path) != token.value()) {
+      LOG_INFO("Access forbidden - invalid HMAC cookie.");
+      sendLocalResponse(403, "", "Access forbidden - invalid HMAC cookie.\n",
+                        {});
+      return FilterHeadersStatus::ContinueAndEndStream;
+    }
+
+    return FilterHeadersStatus::Continue;
+  }
+
+ private:
+  // Try to get the HMAC auth token from the Cookie header.
+  std::optional<std::string> getTokenFromCookie() {
+    const auto cookies = getRequestHeader("Cookie")->toString();
+    std::map<std::string, std::string> m;
+    for (absl::string_view sp : absl::StrSplit(cookies, "; ")) {
+      const std::pair<std::string, std::string> cookie =
+          absl::StrSplit(sp, absl::MaxSplits('=', 1));
+      if (cookie.first == "Authorization") {
+        return cookie.second;
+      }
+    }
+
+    return std::nullopt;
+  }
+
+  // Helper function to convert binary data to a hexadecimal string.
+  std::string toHexString(const unsigned char* data, size_t length) {
```

Comment:
Suggest using absl::BytesToHexString

If that can't be used, suggest passing `data` and `length` as a `string_view`.